app/models/concerns/orchestration/compute.rb

```diff
@@ -102,19 +103,51 @@ def delUserData
       attrs = compute_resource.provided_attributes
-      normalize_addresses if attrs.keys.include?(:mac) or attrs.keys.include?(:ip)
+#      normalize_addresses if attrs.keys.include?(:mac) or attrs.keys.include?(:ip)
+
+      # mac and ip are properties of the NIC, and there may be more than one,
+      # so we need to loop. First store the nics returned from Fog in a local
+      # array so we can delete from it
+      fog_nics = vm.interfaces.dup
 
       attrs.each do |foreman_attr, fog_attr |
-        # we can't ensure uniqueness of #foreman_attr using normal rails validations as that gets in a later step in the process
-        # therefore we must validate its not used already in our db.
-        value = vm.send(fog_attr)
-        value ||= find_address if foreman_attr == :ip
-        self.send("#{foreman_attr}=", value)
-
-        if value.blank? or (other_host = Host.send("find_by_#{foreman_attr}", value))
-          delCompute
-          return failure("#{foreman_attr} #{value} is already used by #{other_host}") if other_host
-          return failure("#{foreman_attr} value is blank!")
+        if foreman_attr == :mac
+          #TODO, do we need handle :ip as well? for openstack / ec2 we only set a single
+          # interface (so host.ip will be fine), and we'd need to rethink #find_address :/
+
+          self.interfaces.each do |nic|
+            selected_nic = vm.select_nic(fog_nics, nic)
+            next if selected_nic.nil? # found no matching fog nic for this Foreman nic, move on
+            value   = selected_nic.send(fog_attr)
+            logger.debug "Orchestration::Compute: nic #{nic.inspect} assigned to #{selected_nic.inspect}"
+            nic.send("#{foreman_attr}=",value)
+            fog_nics.delete(selected_nic) # don't use the same fog nic twice
+
+            # we can't ensure uniqueness of #foreman_attr using normal rails
+            # validations as that gets in a later step in the process
+            # therefore we must validate its not used already in our db.
+
+            # In future, we probably want to skip validation of macs/ips on the Nic
+            # macs can be duplicated if we are creating bonds
+            # ips can be duplicated if we have isolated subnets (needs an update in the Subnet model first)
+            # For now, we scope to physical devices only for the validations
+            if value.blank? or (other = Nic::Base.physical.send("find_by_#{foreman_attr}", value))
+              delCompute
+              return failure("#{foreman_attr} #{value} is already used by #{other}") if other
+              return failure("#{foreman_attr} value is blank!")
+            end
+          end
+        else
+          value = vm.send(fog_attr)
+          value ||= find_address if foreman_attr == :ip
+          self.send("#{foreman_attr}=", value)
+
+          # Check the db uniqueness here too, as per above
+          if value.blank? or (other = Host.send("find_by_#{foreman_attr}", value))
```

Comment:
It's not *quite* duplicated, and I recall that my last attempt to extract it ended in disaster. I'll look again when dealing the previous comment and either do it or detail why it can't be done.